### ISSUE TITLE
Collapse/Group the same kind of spans under a large trace and show count

### ISSUE THREAD

**Issue by srikanthccv** [SigNoz/signoz#5078]
Imagine a trace with the following structure. op1 -> op2 -> op3 -> ... op3 10k times -> op4. In general, for such large traces, there is little to no value in showing all the spans. It would be easy to consume if they are collapsed with total count and duration by default.

**Comment by makeavish** [comment]
@GeekBoySupreme This is planned in new trace detail design, right?

**Comment by GeekBoySupreme** [comment]
Small update. Taking care of this.

**Comment by GeekBoySupreme** [comment]
Does something like this work?

<img width="571" alt="image" src="https://github.com/user-attachments/assets/017a79bb-012e-4134-8bf4-be3d4e7634ec">

**Comment by makeavish** [comment]
> Does something like this work?
> 
> 
> 
> <img width="571" alt="image" src="https://github.com/user-attachments/assets/017a79bb-012e-4134-8bf4-be3d4e7634ec">
> 
> 

LGTM
@srikanthccv What do you think?

**Comment by srikanthccv** [comment]
The two durations are confusing. Can we just show avg duration only?

**Comment by GeekBoySupreme** [comment]
> Can we just show avg duration only?
Do you want to show average duration or the total duration of the entire span group? The first number here - `12ms` denotes the time taken by the total span group. Lmk what we'd want to go ahead with.

**Comment by ankitnayan** [comment]
> In general, for such large traces, there is little to no value in showing all the spans. 

I don't think this is correct, op1 -> op2 -> op3 -> op4 often means cross service client <-> server call and investigating them is very important.

Not sure about the use of showing avg duration of all the children spans combined too. The duration of the parent span which is collapsed should give an idea about the width of time all the children take together.

Can you give me an example from https://stagingapp.signoz.io/trace/00000000000000006e234c6056316e92?spanId=2678eb0c4ed31240&levelUp=0&levelDown=0 @srikanthccv ?

I understand the unnecessary default expanded state of all children which can be changed to collapse all children by default and we already show the count today. Do we really need any proxy for duration of children spans?

**Comment by srikanthccv** [comment]
@ankitnayan taking the same trace you linked, here you can see 10 `HTTP GET /route` under `HTTP GET /dispatch`. Now, make it 1k or 10k `HTTP GET /route` spans which are either exactly or mostly similar except their start and end time. Do I find value in seeing all `HTTP GET /route`, certainly not, such traces exhibit a clear pattern that you can combine them all under one `HTTP GET /route` when possible. This span could be a function that gets invoked hundreds or thousands of times either parallel or serially, I don't see value in showing them all, the only exception is when a few of them out of 10k spans deviate from the rest, in that case, I would find combining those that have same data into the single span and keep the deviated spans unchanged to be most helpful. 
![Screenshot 2024-09-22 at 9 23 05 AM](https://github.com/user-attachments/assets/3f53db4b-604d-456e-9afd-4c4bfab27c24)

**Comment by ankitnayan** [comment]
So, you are talking about flat spans, I confused it with nested spans. The extra grouping might confuse users with the parent-child relationship? If the 5th `HTTP GET /route` has error, you would first expand the group and then goto the error span. 
I got the problem, I have planned something similar in another view called Map where each node will be a unique key based on `serviceName, span name` and the aggregates we show would be analogous to the RED metrics with span count, error spans, avg/min/max latency of spans. Have a look at https://www.notion.so/signoz/Trace-Details-Improvements-730cb90336a04b0e9bc9bcac1677f6ac?pvs=4#109fcc6bcd19808b9985cec6680ad1a1. This is like an aggregate view of a trace and clicking on nodes should be able to take users to the relevant spans in the waterfall tab.

Let me know what you think especially regarding changing the waterfall view to show grouped spans with basic RED stats vs a map.

**Comment by ankitnayan** [comment]
@krsoninikhil PTAL at the problem statement and if this is better solved at spanList view or trace map view instead of waterfall

**Comment by krsoninikhil** [comment]
Discussed, planning to revisit this after span list and trace map feature is developed. Since this might make the pagination logic for large traces a bit complicated and not have any raw view. 

Also, a simple logic based on span name might not be enough to collapse the spans since they can have a different child tree.
